### PR TITLE
`Programming exercises`: Fix an issue that caused tasks to get duplicated during import

### PR DESCRIPTION
**Pull request by Hialus**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1/2 Course(s)

1. Log in to Artemis
2. Check that the existing programming exercise has the correct number of tasks
3. Import the existing programming exercise into the other course
4. Check that the imported programming exercise has the correct number of tasks

To check the number of tasks do one of these steps:
- Check the entries in your local database
- Call the endpoint `/api/programming-exercises/{exerciseId}/tasks` manually
- Apply this patch locally and you will get a button on the PE detail page
```diff
Index: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html b/src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html
--- a/src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html	(revision a01031dd57e0d0257ed6219cd78a9b96ff9f11ff)
+++ b/src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html	(date 1658825544410)
@@ -148,6 +148,16 @@
                             <span jhiTranslate="artemisApp.programmingExercise.structureTestOracle">Update Structure Test Oracle</span>
                         </button>
                     </span>
+                    <span *ngIf="programmingExercise.isAtLeastTutor" class="me-1">
+                        <button
+                            [jhiFeatureToggle]="FeatureToggle.ProgrammingExercises"
+                            type="button"
+                            (click)="getTasks()"
+                            class="btn btn-success btn-sm"
+                        >
+                            <span>Show tasks</span>
+                        </button>
+                    </span>
                 </div>
                 <div *ngIf="programmingExercise.isAtLeastEditor">
                     <span
Index: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts b/src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts
--- a/src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts	(revision a01031dd57e0d0257ed6219cd78a9b96ff9f11ff)
+++ b/src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts	(date 1658825718656)
@@ -453,6 +453,14 @@
         });
     }
 
+    getTasks(): void {
+        if (this.programmingExercise?.id) {
+            this.programmingExerciseService.getTasksAndTestsExtractedFromProblemStatement(this.programmingExercise.id).subscribe((tasks) => {
+                this.alertService.success(`There are ${tasks.length} tasks for this exercise.`);
+            });
+        }
+    }
+
     createBehavioralSolutionEntries() {
         this.programmingExerciseService.createBehavioralSolutionEntries(this.programmingExercise.id!).subscribe({
             next: () => {

```

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Review by canberkanar — DISMISSED** [review]
Code changes look good. 👍🏻

**Review by Strohgelaender — APPROVED** [review]
code 👍

**Review by ole-ve — DISMISSED** [review]
Tested locally and on TS1, both worked as expected with different configurations and for exam exercises.

**Review by ole-ve — APPROVED** [review]
Re-approve manual testing

**Review by nairaarnez — DISMISSED** [review]
Tested in TS1. Looks good!

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseTestService.java [hunk 1]
```diff
@@ -487,6 +504,9 @@ void createAndImportJavaProgrammingExercise(boolean staticCodeAnalysisEnabled) t
                 ProgrammingExercise.class, params, HttpStatus.OK);
         importedExercise = database.loadProgrammingExerciseWithEagerReferences(importedExercise);
 
+        // Check that the tasks were imported correctly (see #5474)
+        assertThat(programmingExerciseTaskRepository.findByExerciseId(importedExercise.getId())).hasSize(sourceExercise.getTasks().size());
```
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseTestService.java:508]:
```suggestion
        assertThat(programmingExerciseTaskRepository.findByExerciseId(importedExercise.getId())).hasSameSizeAs(sourceExercise.getTasks());
```

**Review by ole-ve — APPROVED** [review]
Manual testing

**Review by Strohgelaender — APPROVED** [review]

**Review by bassner — APPROVED** [review]
Tested on TS1 and worked fine - a default BubbleSort exercise has been created that had 7 tasks. Imported it afterwards, the imported clone had 7 tasks as well.

Code LGTM 👍

**Review by canberkanar — APPROVED** [review]
Reapprove 👍🏻

**Review by babyygemperor — APPROVED** [review]
Tested on ts1, works as described

**Review by krusche — APPROVED** [review]
The changes fix the issue, however I suggest to avoid the code duplication by moving the new 2nd query into the method `importProgrammingExerciseBasis` or even into the method `importTasks`

**Pull request merged**